man/kcp.1
.TH "kcp" 1 "2015-08-27" "kcp 0.80" "User Manuals"

.SH NAME
kcp \- Tool in command-line for KaOS Community Packages

.SH SYNOPSIS
kcp (\-hvlusgicxfNSIOdDV) [<app>]

.SH DESCRIPTION
kcp provides a tool to make the use of KaOS Community Packages easier. With this tool, you can:

- list all packaes of KaOS Community Packages,

- search a package in KaOS Community Packages,

- download needed files to install a package from KaOS Community Packages,

- install a package of KaOS Community Packages.

.SH COMMAND OPTIONS

.TP
.B \-h, \-\-help
Display the list of the available options

.TP
.B \-v, \-\-version
Display the current version of kcp

.TP
.B \-u, \-\-update-database
Refresh the local database. The database is a file (see CREATED FILES chapter) which contains the following informations:

- The name of a package,

- Its description,

- Its installed version (if installed),

- Its latest version in KaOS Community Packages,

- Its popularity.

.TP
.B \-l, \-\-list
Display the list of all packages. By default, it displays all informations contained in the database (name, remote/local version, popularity & description)

.TP
.B \-s, \-\-search <app>
Display the list of all packages which contain the keyword <app> in the name or the description.

.TP
.B \-g, \-\-get <app>
Download the package <app> from KaOS Community Packages in the current directory.

.TP
.B \-i, \-\-install <app>
Download, compile and install the package <app> from KaOS Community Packages.

.TP
.B \-V, \-\-information <app>
Display information on a given package.


.SH SPECIFIC OPTIONS

.TP
.B \-f, \-\-force-update
By default, displaying list of packages uses the informations from the local database. With this option, the database is refreshed before displaying the list. It can be used only if -l or -s option is used.

.TP
.B \-x, \-\-sort
On packages' display operation, sort the results by popularity, then, by name. This option can be used only with -l or -s options.

.TP
.B \-N, \-\-only-name
On packages' display operation, display only the name of the packages. This option can be used only with -l or -s options.

.TP
.B \-S, \-\-only-starred
On packages' display operation, display only the packages which are popular. This option can be used only with -l or -s options.

.TP
.B \-I, \-\-only-installed
On packages' display operation, display only the packages which are installed on the system. This option can be used only with -l or -s options.

.TP
.B \-O, \-\-only-outdated
On packages' display operation, display only the packages which are installed on the system and whose installed version isn't the same as KaOS Community Packages version. This option can be used only with -l or -s options.

.TP
.B \-d, \-\-asdeps
Install packages non-explicitly; in other words, fake their install reason to be installed as a dependency. This is useful to install dependencies before building the package.

.TP
.B \-D, \-\-debug
For internal use only. Display the header of the requests' responses, in order to identify a potential problem.

.SH CREATED FILES

.TP
~/.kcp.json
This file is the local database. It is created at first 'kcp -l' or 'kcp -u' command and refresh each time -u or -f options are used.

.TP
/tmp/kcp.lock
This file is created on package installation from KaOS Community Repositories. It is used to prevent usage of two instances of kcp. When installing a package <app>, a temporary <app> directory is created to download the needed files for compilation. Both are removed automatically at end of command (even on prematured exit).

.SH CREDITS

.TP
This is free and unencumbered software released into the public domain.
http://unlicense.org/ for more information.

.SH SEE ALSO

For more information regarding kcp, see the following:

.TP
The official code repository at
<git://github.com/bvaudour/kcp>

.TP
You can contact the project by emailing
<benjamin.vaudour@gmx.fr>

.SH REPORTING BUGS
.TP
Bugs in kcp may be reported to the issue-tracker at
<https://github.com/bvaudour/kcp/issues>


kcp.py
#!/usr/bin/env python3
# License: Public Domain
# Release: 0.8

import argparse, os, sys, subprocess, json
from urllib import request

search_head  = 'application/vnd.github.v3.text-match+json'
search_h_tp  = 'Accept'
search_base  = 'https://api.github.com/search/repositories?q={}+user:KaOS-Community-Packages'
url_base     = 'https://github.com/KaOS-Community-Packages/{}.git'
url_pkgbuild = 'https://raw.githubusercontent.com/KaOS-Community-Packages/{}/master/PKGBUILD'

def print_error(msg):
	print('\033[1;31m{}\033[m'.format(msg))

def question(msg, default_value=True):
	if default_value:
		choice_rep = '[Y/n]'
	else:
		choice_rep = '[y/N]'
	response = input('\033[1;33m{} {} \033[m'.format(msg, choice_rep)).lower()
	if not response:
		return default_value
	if response[0] == 'y':
		return True
	if response[0] == 'n':
		return False
	return default_value

def edit(file_name):
	os.system('$EDITOR {}'.format(file_name))

def check_user():
	if not os.geteuid():
		print_error("Don't launch this program as root!")
		sys.exit(1)

def edit_pkgbuild():
	if question('Do you want to edit PKGBUILD?'):
		edit('PKGBUILD')

def get_package(app):
	url = url_base.format(app)
	exe = subprocess.Popen(['git', 'clone', url])
	err = exe.wait()
	if err:
		sys.exit(err)

def launch_request(search, header = None):
	req = request.Request(search)
	if header:
		req.add_header(search_h_tp, header)
	try:
		result = request.urlopen(req)
		return result.read().decode()
	except:
		return ''

def get_version(result):
	pkgver, pkgrel = '', ''
	for l in result.split('\n'):
		e = l.strip()
		if e[:7] == 'pkgver=':
			pkgver = e[7:]
		elif e[:7] == 'pkgrel=':
			pkgrel = e[7:]
	if pkgver and pkgrel:
		return '{}-{}'.format(pkgver, pkgrel)
	return '<unknown>'

def check_installed(app):
	exe = subprocess.Popen(['pacman', '-Q', app], stdout=subprocess.PIPE, stderr=subprocess.PIPE)	
	if exe.wait():
		return ''
	return exe.stdout.read().decode().split()[1]

def search_package(app, fast):
	search = search_base.format(app)
	result = json.loads(launch_request(search, search_head))
	for a in result['items']:
		n, d, s = a['name'], a['description'], a['stargazers_count']
		i = check_installed(a['name'])
		if fast:
			if i:
				print('\033[1m{}\033[m \033[1;36m[installed: {}]\033[m \033[1;34m({})\033[m'.format(n, i, s))
			else:
				print('\033[1m{}\033[m\033 \033[1;34m({})\033[m'.format(n, s))
		else:
			v = get_version(launch_request(url_pkgbuild.format(n)))
			if i:
				if v == i:
					i = ' [installed]'
				else:
					i = ' [installed: {}]'.format(i)
			print('\033[1m{}\033[m \033[1;32m{}\033[m\033[1;36m{}\033[m \033[1;34m({})\033[m'.format(n, v, i, s))
		print('\t{}'.format(d))

def install_package(app, asdeps):
	os.chdir('/tmp')
	get_package(app)
	os.chdir('/tmp/{}'.format(app))
	edit_pkgbuild()
	cmd = 'makepkg -si'
	if asdeps:
		cmd += ' --asdeps'
	os.system(cmd)
	os.system('rm -rf {}'.format(os.getcwd()))

def build_args():
	parser = argparse.ArgumentParser(description='Tool in command-line for KaOS Community Packages')
	parser.add_argument('-v', '--version', help='print version', action='version', version='0.8')
	group = parser.add_mutually_exclusive_group()
	group.add_argument('-g', '--get', help='get needed files to build app', metavar='APP')
	group.add_argument('-s', '--search', help='search an app in KCP', metavar='APP')
	group.add_argument('-i', '--install', help='install an app in KCP', metavar='APP')
	parser.add_argument('--asdeps', help='install as a dependence', action='store_true', default=False)
	parser.add_argument('--fast', help='search without version', action='store_true', default=False)
	return parser

check_user()

parser = build_args()
if len(sys.argv) == 1:
	parser.print_help()
	sys.exit(0)

args = parser.parse_args()
if args.get:
	get_package(args.get)
elif args.search:
	search_package(args.search, args.fast)
elif args.install:
	install_package(args.install, args.asdeps)
else:
	parser.print_help()
